### DISCUSSION TITLE
How to create nested components from expressions?

### DISCUSSION THREAD

**Discussion by hlvlad** [molstar/molstar#398] [Q&A]
Hello,

Given following simplified code:

```typescript
const residues = [{chainId: 'A', residueId: 1}, {chainId:'B', residueId: 2}];

const residueExpressions = residues.map(residue => 
    MS.core.logic.and([
        MS.core.rel.eq([residue.chainId, MS.ammp("auth_asym_id")]),
        MS.core.rel.eq([residue.residueId, MS.ammp("auth_seq_id")])
    ])
);
const residuePairExpression = MS.core.logic.or(residueExpressions);
const residuePairComponent = await plugin.builders.structure.tryCreateComponentFromExpression(
    structure,
    MS.struct.generator.atomGroups({
      "residue-test": residuePairExpression
    }),
    "residue-pair",
    {label: "Residue Pair"}
);
if (!residuePairComponent) return;
for(let i = 0; i < 2; i++) {
   const residueComponent = await plugin.builders.structure.tryCreateComponentFromExpression(
      residuePairComponent?.ref,
      residueExpressions[i],
      `residue-${i}`,
      {label: `Residue ${i}`}
    );
    if (residueComponent) await plugin.builders.structure.representation.addRepresentation(residueComponent, {
       type: "cartoon",
    });
}
const otherResiduesComponent = await plugin.builders.structure.tryCreateComponentFromExpression(
    structure,
    MS.struct.generator.atomGroups({
      "residue-test": MS.core.logic.not([residuePairExpression]);
    }),
    "other-residues",
    {label: "Other Residues"}
);
```

I have few questions:
1. Is there more simple way to create identical state tree, without first querying original structure, and then querying child component (`residue-pair`)? Somehow create empty `residue-pair` and then query its children from structure?
2. Is there more idiomatic way to create `other-residues` (find all residues not in `residue-pair` by set complement)?

**Comment by hlvlad** [comment]
@arose, @dsehnal could you give advice on this? Thank you in advance.

**Reply by dsehnal** [reply]
You can run the queries outside the state tree and then just create the component by converting the final result to a "bundle" and creating it from that. 

- Create: ``StructureQueryHelper.createAndRun``
- Converts to structure to run again: ``StructureSelection.unionStructure``
- Bundle: ``StructureSelection.unionStructure`` + ``StructureElement.Bundle``
